File: Strivers-A2Z-DSA-Sheet/13. Graphs/3. Topo Sort Problems/python/05. Find Eventual Safe State.py
"""
Question:
Find all safe nodes in a directed graph. A node is safe if every possible path
starting from that node leads to a terminal node.

Approach:
- Build reverse graph and track outdegree of each node.
- Start BFS from terminal nodes (outdegree 0).
- Nodes that become terminal during BFS are safe nodes.

Time Complexity: O(N + E)
Space Complexity: O(N + E)
"""

from collections import deque


def eventual_safe_nodes(graph):
    n = len(graph)
    revadj = [[] for _ in range(n)]
    outdeg = [0] * n

    for i in range(n):
        outdeg[i] = len(graph[i])
        for j in graph[i]:
            revadj[j].append(i)

    queue = deque()
    for i in range(n):
        if outdeg[i] == 0:
            queue.append(i)

    ans = []
    while queue:
        node = queue.popleft()
        ans.append(node)
        for v in revadj[node]:
            outdeg[v] -= 1
            if outdeg[v] == 0:
                queue.append(v)

    ans.sort()
    return ans

File: Strivers-A2Z-DSA-Sheet/13. Graphs/3. Topo Sort Problems/javascript/05. Find Eventual Safe State.js
/*
Question:
Find all safe nodes in a directed graph. A node is safe if every possible path
starting from that node leads to a terminal node.

Approach:
- Build reverse graph and track outdegree of each node.
- Start BFS from terminal nodes (outdegree 0).
- Nodes that become terminal during BFS are safe nodes.

Time Complexity: O(N + E)
Space Complexity: O(N + E)
*/

function eventualSafeNodes(graph) {
    let n = graph.length;
    let revadj = Array.from({ length: n }, () => []);
    let outdeg = new Array(n).fill(0);

    for (let i = 0; i < n; i++) {
        outdeg[i] = graph[i].length;
        for (let j of graph[i]) {
            revadj[j].push(i);
        }
    }

    let queue = [];
    for (let i = 0; i < n; i++) {
        if (outdeg[i] === 0) queue.push(i);
    }

    let ans = [];
    while (queue.length > 0) {
        let next = [];
        for (let node of queue) {
            ans.push(node);
            for (let v of revadj[node]) {
                outdeg[v]--;
                if (outdeg[v] === 0) next.push(v);
            }
        }
        queue = next;
    }

    ans.sort((a, b) => a - b);
    return ans;
}

